RHIDP-4807: Third-party plugin installation in RHDH (#733)
redhat-developer/red-hat-developers-documentation-rhdh: assemblies/dynamic-plugins/assembly-third-party-plugins-installation.adoc

[id="assembly-third-party-plugins-installation"]
= Third-party plugins in {product}
:context: assembly-third-party-plugins-installation

> In {product}, third-party dynamic plugins can be integrated seamlessly, enabling the use of external plugins alongside native ones. This flexibility allows you to enhance the platform functionality without modifying its core structure. To add third-party dynamic plugins, you can either export them as standalone packages or wrap them in custom wrappers, especially if you want to adjust the dependencies or configurations of a plugin for compatibility.

Reviewer: I would also suggest instead of "modifying its core structure" we use something like "modifying its source code and rebuilding".

> In {product}, third-party dynamic plugins can be integrated seamlessly, enabling the use of external plugins alongside native ones. This flexibility allows you to enhance the platform functionality without modifying its core structure. To add third-party dynamic plugins, you can either export them as standalone packages or wrap them in custom wrappers, especially if you want to adjust the dependencies or configurations of a plugin for compatibility.

Author: Done

> The third-party plugins are typically packaged as dynamic plugins, which means they can be loaded at runtime, enabling you to easily manage external modules without rebuilding the entire application. The process involves creating a wrapper that exports the plugin code, ensuring that dependencies are correctly bundled or marked as shared, depending on their relationship to the {product-short} environment.

Author: Done

To integrate a third-party plugin into {product-short}:

. First, obtain the plugin's source code.
. Export the plugin as a dynamic plugin package. See xref:proc-export-third-party-plugins-rhdh_{context}[].
. Package and publish the dynamic plugin. See xref:proc-package-third-party-dynamic-plugin_{context}[].
. Install the plugin in the {product-short} environment. See xref:proc-install-third-party-plugins-rhdh_{context}[].

The third-party plugins include:

> Backend plugins::

Author: Done

+
--

> To ensure compatibility with the dynamic plugin support and enable their use as dynamic plugins, existing backend plugins must be compatible with, the new backend system. Additionally, these plugins must be rebuilt using a dedicated CLI command.

Reviewer: ... must be compatible with, the new...  <-- this comma is unnecessary

> The new backend system is created using `createBackendPlugin()` or `createBackendModule()`. You must export the new backend system as the default export from either the main package or an alpha package (if support exists in alpha). The dynamic export mechanism automatically identifies private dependencies, and adds them to the `bundleDependencies` field in the `package.json` file. This mechanism ensures that the dynamic plugin package is self-contained, with private dependencies bundled within a private `node_modules` folder.

Reviewer: I think in these sentences the term "new backend system" isn't quite accurate.  I think you're talking about the object returned from these APIs which would be the plugin or module.  I wonder though how much of these we need to include here vs rely on the upstream documentation.

Author: Replaced the term "new backend system" with "plugin instance"

> The new backend system is created using `createBackendPlugin()` or `createBackendModule()`. You must export the new backend system as the default export from either the main package or an alpha package (if support exists in alpha). The dynamic export mechanism automatically identifies private dependencies, and adds them to the `bundleDependencies` field in the `package.json` file. This mechanism ensures that the dynamic plugin package is self-contained, with private dependencies bundled within a private `node_modules` folder.

Reviewer: "new backend system instance" or just replace "new backend system" with "plugin instance" perhaps?

The backend plugins contains two types of dependencies, including:

> * *Shared dependencies* are provided by an application and listed as `peerDependencies` in `package.json` file, not bundled in the dynamic plugin package. For example, by default, all `@backstage` packages are shared.

Reviewer: instead of "an application" we could probably instead refer to the main RHDH application here.  The term "an application" is better used when this functionality is made portable and installable in any backstage, but for probably for the product docs referring to the RHDH application here is better.

+
You can use the `--shared-package` flag to specify shared dependencies. To treat a `@backstage` package as private, use the negation prefix (`!`).

* *Embedded dependencies* are bundled into the dynamic plugin package with their dependencies hoisted to the top level. By default, packages with `-node` or `-common` suffixes are embedded.
+
You can use the `--embed-package` flag to specify additional embedded packages, including those in the same repository that do not follow the default naming convention.

The following is an example of exporting a dynamic plugin with shared and embedded packages:

.Example dynamic plugin export with shared and embedded packages
[source,bash]
----
npx @janus-idp/cli@latest export-dynamic-plugin --shared-package '!/@backstage/plugin-notifications/'<1> --embed-package @backstage/plugin-notifications-backend <2>
----

<1> `@backstage/plugin-notifications` package is treated as a private dependency and is bundled in the dynamic plugin package, despite being in the `@backstage` scope.
<2> `@backstage/plugin-notifications-backend` package is marked as an embedded dependency and is bundled in the dynamic plugin package.
--

Front-end plugins::
+
--

> Front-end plugins can use `scalprum` for configuration, which the CLI can generate automatically during the export process. The generated default configuration is logged when running the following command:

Reviewer: Shouldn't this section be in the document about exporting dynamic plugins?  It's not clear to me how this relates to installing third party plugins.

Author: Moved this content to export section

[id="assembly-third-party-plugins-installation"]
= Third-party plugins in {product}
:context: assembly-third-party-plugins-installation

> In {product}, third-party dynamic plugins can be integrated seamlessly, enabling the use of external plugins. This flexibility allows you to enhance the platform functionality without modifying its source code and rebuilding. To add third-party dynamic plugins, you need to export them as derived package.

Reviewer 2: Consider removing fluff, and using active voice.

```suggestion
You can integrate third-party dynamic plugins in {product}. This flexibility allows you to enhance the platform functionality without modifying its source code and rebuilding. To add third-party dynamic plugins, export them as derived package.
```

> The third-party plugins are typically packaged as dynamic plugin derived package, enabling you to manage external modules without rebuilding the entire application. The process involves exporting the plugin package, ensuring that dependencies are correctly bundled or marked as shared, depending on their relationship to the {product-short} environment.

Reviewer 2: Consider removing repeated content. Note that this sentence is still a definition of the step in line 12.

```suggestion
While exporting the plugin package, you should ensure that dependencies are correctly bundled or marked as shared, depending on their relationship to the {product-short} environment.
```
